File: Source/FSharp.Example.Streams.Chat.Server/ChatUser.fs
namespace Actors

open FSharp.Control.Tasks

open Orleankka
open Orleankka.FSharp

open Messages

type ChatUser() =
    inherit ActorGrain()

    member this.send roomName message = 
        printfn "[server]: %s" message
        let room = ActorSystem.streamOf(this.System, "sms", roomName)
        room <! { UserName = this.Id; Text = message } 
    
    interface IChatUser
    override this.Receive(message) = task {      
        match message with
        | :? ChatUserMessage as m -> 
            match m with
            | Join room         ->  let msg = sprintf "%s joined the room %s ..." this.Id room
                                    do! this.send room msg
                                    return none()    
                          
            | Leave room        ->  let msg = sprintf "%s left the room %s!" this.Id room
                                    do! this.send room msg
                                    return none()

            | Say (room,msg)    ->  let msg = sprintf "%s said: %s" this.Id msg
                                    do! this.send room msg
                                    return none()

        | _ -> return unhandled()
   }

File: Samples/CSharp/Streams/Chat.Server/ChatUser.cs
using System;
using System.Threading.Tasks;

using Orleankka;

namespace Example
{
    public class ChatUser : DispatchActorGrain, IChatUser
    {
        Task On(Join x)   => Send(x.Room, $"{Id} joined the room {x.Room} ...");
        Task On(Leave x)  => Send(x.Room, $"{Id} left the room {x.Room}!");
        Task On(Say x)    => Send(x.Room, $"{Id} said: {x.Message}");

        Task Send(string room, string message)
        {
            Console.WriteLine("[server]: " + message);

            var stream = System.StreamOf("sms", room);

            return stream.Publish(new ChatRoomMessage
            {
                User = Id,
                Text = message
            });
        }
    }
}